File: Grind75/12-LinkedListCycle/cpp/LinkedListCycle.cpp
#include "LinkedListCycle.h"
#include <unordered_set>

bool LinkedListCycle::hasCycle(List::ListNode *head) {
  List::ListNode *slow = head;
  List::ListNode *fast = (head) ? head->next : nullptr;

  while (slow && fast) {
    if (slow == fast)
      return true;
    slow = slow->next;
    fast = fast->next;
    if (fast)
      fast = fast->next;
  }
  return false;
}

bool LinkedListCycle::hasCycle_hashset(List::ListNode *head) {
  List::ListNode *node = head;
  std::unordered_set<List::ListNode*> visited;
  while (node) {
    if (visited.find(node) != visited.end())
      return true;
    visited.insert(node);
    node = node->next;
  }
  return false;
}

File: Grind75/12-LinkedListCycle/python/LinkedListCycle.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from List import *
from typing import Optional

class Solution:
  def hasCycle(self, head: Optional[ListNode]) -> bool:
    if head is None:
      return False
    slow = head
    fast = head.next
    while fast is not None:
      if slow is fast:
        return True
      slow = slow.next
      fast = fast.next
      if fast is None:
        break
      fast = fast.next
    return False

if __name__ == '__main__':
  soln = Solution()
  testCases = []

  lst = makeList([1,2,3])
  print(soln.hasCycle(lst))
  lst.next.next.next = lst
  print(soln.hasCycle(lst))
  
  for test in testCases:
    pass
  #print('TESTS PASSED!')

